src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```diff
@@ -0,0 +1,361 @@
+                            // |VER | ULEN |  UNAME   | PLEN |  PASSWD  |
+                            // +----+------+----------+------+----------+
+                            // | 1  |  1   | 1 to 255 |  1   | 1 to 255 |
+                            // +----+------+----------+------+----------+
+                            buffer[0] = ProtocolVersion5;
+                            int uLen = Encoding.UTF8.GetByteCount(credentials.UserName);
+                            buffer[1] = checked((byte)uLen);
+                            int uLenEncoded = Encoding.UTF8.GetBytes(credentials.UserName, buffer.AsSpan(2));
+                            Debug.Assert(uLen == uLenEncoded);
+                            int pLen = Encoding.UTF8.GetByteCount(credentials.Password);
+                            buffer[2 + uLen] = checked((byte)pLen);
+                            int pLenEncoded = Encoding.UTF8.GetBytes(credentials.Password, buffer.AsSpan(3 + uLen));
+                            Debug.Assert(pLen == pLenEncoded);
+                            await WriteAsync(stream, buffer.AsMemory(0, 4 + uLen + pLen), async).ConfigureAwait(false);
+
+                            // +----+--------+
+                            // |VER | STATUS |
+                            // +----+--------+
+                            // | 1  |   1    |
+                            // +----+--------+
+                            await ReadToFillAsync(stream, buffer.AsMemory(0, 2), async).ConfigureAwait(false);
+                            if (buffer[0] != ProtocolVersion5)
+                                throw new Exception("Bad protocol version");
+                            if (buffer[1] != REP_SUCCESS)
+                                throw new Exception("Authentication failed.");
+                            break;
+                        }
+
+                    case METHOD_NO_ACCEPTABLE:
+                    default:
+                        throw new Exception("No acceptable auth method.");
+                }
+
+
+                // +----+-----+-------+------+----------+----------+
+                // |VER | CMD |  RSV  | ATYP | DST.ADDR | DST.PORT |
+                // +----+-----+-------+------+----------+----------+
+                // | 1  |  1  | X'00' |  1   | Variable |    2     |
+                // +----+-----+-------+------+----------+----------+
+                buffer[0] = ProtocolVersion5;
+                buffer[1] = CMD_CONNECT;
+                buffer[2] = 0;
+                int addressLength;
+
+                if (IPAddress.TryParse(host, out var hostIP))
+                {
+                    if (hostIP.AddressFamily == Sockets.AddressFamily.InterNetwork)
+                    {
+                        buffer[3] = ATYP_IPV4;
+                        hostIP.TryWriteBytes(buffer.AsSpan(4), out int bytesWritten);
+                        Debug.Assert(bytesWritten == 4);
+                        addressLength = 3;
```

Comment:
This looks confusing. We can use `addressLength = 4`, `addressLength = 16` instead and increment the value for the domain-name case